tensorflow_lite_support/c/task/vision/image_classifier.cc

```diff
@@ -0,0 +1,225 @@
+using ClassificationsCpp = ::tflite::task::vision::Classifications;
+using ClassCpp = ::tflite::task::vision::Class;
+using BoundingBoxCpp = ::tflite::task::vision::BoundingBox;
+using ImageClassifierCpp = ::tflite::task::vision::ImageClassifier;
+using ImageClassifierOptionsCpp =
+    ::tflite::task::vision::ImageClassifierOptions;
+using FrameBufferCpp = ::tflite::task::vision::FrameBuffer;
+}  // namespace
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+struct TfLiteImageClassifier {
+  std::unique_ptr<ImageClassifierCpp> impl;
+};
+
+std::unique_ptr<ImageClassifierOptionsCpp>
+CreateImageClassifierCppOptionsFromCOptions(
+    const TfLiteImageClassifierOptions* c_options) {
+  std::unique_ptr<ImageClassifierOptionsCpp> cpp_options(
+      new ImageClassifierOptionsCpp);
+
+  // More file sources can be added in else ifs
+  if (c_options->base_options.model_file.file_path)
+    cpp_options->mutable_base_options()->mutable_model_file()->set_file_name(
+        c_options->base_options.model_file.file_path);
+  else
+    return nullptr;
+
+  if (c_options->base_options.compute_settings
+                               .tflite_settings.cpu_settings.num_threads > 0)
+              cpp_options->mutable_base_options()
+          ->mutable_compute_settings()
+          ->mutable_tflite_settings()
+          ->mutable_cpu_settings()
+          ->set_num_threads(c_options->base_options.compute_settings
+                                .tflite_settings.cpu_settings.num_threads);
+  else {
+    cpp_options->mutable_base_options()
+        ->mutable_compute_settings()
+        ->mutable_tflite_settings()
+        ->mutable_cpu_settings()
+        ->set_num_threads(-1);
+  }
+
+  if (c_options->classification_options.class_name_blacklist.length > 0 &&
+      c_options->classification_options.class_name_whitelist.length > 0)
+    return nullptr;
+
+  if (c_options->classification_options.class_name_blacklist.length > 0 &&
+      c_options->classification_options.class_name_whitelist.length)
+    return nullptr;
+  
+  if (c_options->classification_options.class_name_blacklist.length > 0) {
+    for (int i = 0;
+         i < c_options->classification_options.class_name_blacklist.length; i++)
+      cpp_options->add_class_name_blacklist(
+          c_options->classification_options.class_name_blacklist.list[i]);
+  }
+  
+  if (c_options->classification_options.class_name_whitelist.length >
+             0) {
+    for (int i = 0;
+         i < c_options->classification_options.class_name_whitelist.length; i++)
+      cpp_options->add_class_name_whitelist(
+          c_options->classification_options.class_name_whitelist.list[i]);
+  }
+
+  if (c_options->classification_options.display_names_local) {
+    cpp_options->set_display_names_locale(
+        c_options->classification_options.display_names_local);
+  }
+
+  if (c_options->classification_options.max_results > 0) {
```

Comment:
If check can be removed. Just rely on the  C++ layer to check.

Reply:
We can't remove this check. The C `TfLiteImageClassifierOptions` needs to be zero initialised by users of C api to prevent garbage values in members. Otherwise it will result in seg faults in C and C++.  I have mentioned this in the documentation. `max_results` gets initialised to 0 by default in this case. Setting 0 for `max_results` in cpp layer results in a not ok status for creation of classifier. This may not be what the user of the C API intended to do. 

Ideally from C, you should be able to set only options you need and the underlying layers should take care of initialising others to default values. IMO the CPP layer should use default `max_results` when it is 0. Until then we will need this check given we can't have any defaults for struct members in C. 

Another workaround is to have a function in the C layer to return a default `TfLiteImageClassifierOptions` set with defaults and mandate that C options should be created only using this function. IMO this needs to be done. Then the programmer can set the options to whatever values he wants and also not worry about zero initialisation. In this case we can avoid these checks completely since the onus is on them to correctly change values.

Same applies for `num_threads`. We can remove the check for `score_threshold` since 0 is the expected default. Garbage values shouldn't be the frameworks headache.

Reply:
score_threshold is a bit different. If you set 0, then everything below 0 will be filtered. See the code [here](  https://github.com/tensorflow/tflite-support/blob/17cdc506a620e3a0745a11023ebc2542a5237d25/tensorflow_lite_support/cc/task/processor/classification_postprocessor.cc#L205). How about setting the default to FLT_MAX? See https://en.cppreference.com/w/c/types/limits.

Reply:
Score threshold default should be 0 right? FLT_MAX would filter everything out as score is probabilities. By default everything should be returned IMO.
max_results default can be  -1. [Doc](https://github.com/tensorflow/tflite-support/blob/master/tensorflow_lite_support/cc/task/processor/proto/classification_options.proto#L27-L30) says so.